### PR TITLE
Mesh constraint matrix support in FileMeshGenerator

### PR DESCRIPTION
**Pull request by roystgnr**
## Reason
This lets us combine an arbitrary extraction operator (expressed as a sparse matrix file) with a mesh of assembly elements, in a generalization of what we do with IsoGeometric Analysis meshes now that lets us support more complicated mesh types.

## Design
If a constraint matrix file is requested, we load it and import its constraints along with the mesh file it applies to; doing this in FileMeshGenerator rather than a separate generator ensures we can control renumbering settings and make sure the ordering from the files stays matched.

## Impact
No impact for users who don't enable the option; more mesh types available for users who do.

Oh, except I also slipped in a bug-fix/optimization that I noticed while setting up a regression test using SideSetsFromNormalsGenerator

Refs #18768

### CODE REVIEW

**Comment by roystgnr** [comment]
Those remaining failures look real - I'm seeing different exodiff results on different processor counts, larger than roundoff error could explain - but I'm not able to reproduce them in the read/constrain/solve unit tests in https://github.com/libMesh/libmesh/pull/3831, so I'd wondering if I'm doing something wrong in our ExodusII output, but in paraview the diffs look too smooth to just be some sort of straightforward misindexing issue.

Looks like I can reproduce the problem in 1D, which should be simple enough to debug by hand.

**Review by GiudGiud — COMMENTED** [review]
I would create a parameter group for the IGA parameters now since there are a few with
```
params.addParamNameToGroup(params, "group")
```

A bit of doco in FileMeshGenerator.md would help too?

File: framework/src/meshgenerators/FileMeshGenerator.C [hunk 1]
```diff
@@ -70,6 +77,14 @@ FileMeshGenerator::generate()
 {
   auto mesh = buildMeshBaseObject();
 
+  // Maybe we'll reallow renumbering after constraints are applied?
+  bool eventually_allow_renumbering = _allow_renumbering && mesh->allow_renumbering();
+
+  // If we have a constraint matrix, we need its numbering to match
+  // the numbering in the mesh file
+  if (_matrix_file_name != "")
```
Comment (lindsayad) [inline framework/src/meshgenerators/FileMeshGenerator.C:85]:
```suggestion
  if (!_matrix_file_name.empty())
```

File: framework/src/meshgenerators/FileMeshGenerator.C [hunk 2]
```diff
@@ -129,6 +145,25 @@ FileMeshGenerator::generate()
     _app.possiblyLoadRestartableMetaData(MooseApp::MESH_META_DATA, (std::string)file_name);
   }
 
+  if (_matrix_file_name != "")
```
Comment (lindsayad) [inline framework/src/meshgenerators/FileMeshGenerator.C:148]:
```suggestion
  if (!_matrix_file_name.empty())
```

File: test/tests/meshgenerators/file_mesh_generator/tests [hunk 3]
```diff
@@ -157,6 +157,38 @@
     issues = '#18768'
   []
 
+  [./iga_constraint_matrix]
```
Comment (lindsayad) [inline test/tests/meshgenerators/file_mesh_generator/tests:160]:
```suggestion
  [iga_constraint_matrix]
```

File: test/tests/meshgenerators/file_mesh_generator/tests [hunk 4]
```diff
@@ -157,6 +157,38 @@
...
+
+    # Penalty method boundary conditions
+    abs_zero = 1e-08
+
+    requirement = 'The system shall have the ability to apply externally defined constraint matrices to 3D isoparametric meshes.'
+    design = 'meshgenerators/FileMeshGenerator.md'
+    issues = '#18768'
+  []
+
+  [./iga_constraint_matrix_1d]
```
Comment (lindsayad) [inline test/tests/meshgenerators/file_mesh_generator/tests:176]:
```suggestion
  [iga_constraint_matrix_1d]
```

**Review by lindsayad — APPROVED** [review]

**Pull request merged**